mitodl/ocw-studio
# Customizable URLs for studio

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes #1336

#### What's this PR do?
- Adds a `url_path` field to `Website`.  The migration should prepopulate it with the same value as `Website.name` for all sites that have already been published to production.
- When publishing to draft/live, `url_path` can be sent in the request body to set the field's value.  The serializer should allow the field value to change as long as the site hasn't been published to production yet and the value is valid.  If the value is changed, all `WebsiteContent` objects with populated `file` values for the site will have their synced_checksums reset so they will be resynced to github with the updated path.
- Updates the HugoMarkdownSerializer to save the correct `file` values (based on `url_path`) to the backend (git) and vice versa.
- Uses a new `site-url-format` field in site config to suggest a pattern for this field to the user on the front end.
- Changes the publish drawer form, to include a url field if the site has not been published to production yet.
- Updates both individual and mass pipeline definitions so that S3 files are copied to the right places.  Creates/updates the site pipeline before each publish event to ensure that it has the correct url path baked in.

#### How should this be manually tested?
- Update the ocw-course starter config with the [new line here](https://github.com/mitodl/ocw-hugo-projects/pull/180/files).
- Start studio with concourse `docker-compose --profile concourse up`
- Run `manage.py upsert_mass_build_pipeline` and `backpopulate_pipelines`, they should both succeed.
- Try to publish an existing site with a `publish_date`, the drawer should look about the same and work the same.
- Try to publish a new or existing site without a `publish_date`, you should see a URL field with a suggested path, any existing metadata values matching the pattern should be present.  Enter a valid value and click 'Publish' to draft, check to see that the `url_path` is updated and the pipeline is triggered (it will fail, that's ok).
- Go to the github repo and check that any resources for images, pdfs, etc have a `file` value that contains the site's url path.
- Reload the page and try to publish again, the URL field should be set to whatever you entered last time instead of the suggested path.  Modify it and click the "Publish" button, the pipeline should trigger again.  Also check on the git repo and make sure the resource file paths have been updated again to the latest value.
- The changes here affect most parts of the site, so try just about everything and make sure it still works.
- The changes to the pipelines are best tested when this hits RC.

#### Where should the reviewer start?
https://github.com/mitodl/hq/discussions/1


#### Screenshots (if appropriate)
An unpublished site without any user-supplied url path so far:

<img width="734" alt="Screen Shot 2022-05-17 at 3 45 16 PM" src="https://user-images.githubusercontent.com/187676/168899681-4f21e77d-aafe-4956-adb2-39ca550ea4cc.png">

An unpublished site with a previously chosen url (still editable):

<img width="980" alt="Screen Shot 2022-05-17 at 3 53 26 PM" src="https://user-images.githubusercontent.com/187676/168899757-88393b27-19b7-45cd-b63a-bf5bb90dca90.png">

A published site (no longer editable):

<img width="981" alt="Screen Shot 2022-05-17 at 3 54 10 PM" src="https://user-images.githubusercontent.com/187676/168899823-8fb139cf-fd60-4af0-828e-ebebf45b99a2.png">
